### PR TITLE
Allow adding domain mapping via url using composite checkout

### PR DESCRIPTION
**Pull request by sirbrillig**
This adds support for adding products of the form `domain-mapping:example.com` directly from the URL. Note that `domain-mapping` is not a real product slug, but rather a special alias used by calypso URLs for the product `domain_map` (another already-supported example is the alias `premium` for `value_bundle`).

Regular checkout already supports this feature, and with this change, composite checkout will also support it. So that it gets used, this change also modifies `CheckoutSystemDecider` so that it routes requests like `/checkout/YOURSITEHERE/domain-mapping:example.com` to composite checkout.

#### Testing instructions

- Make sure your cart in calypso is empty.
- Visit http://calypso.localhost:3000/checkout/YOURSITEHERE/premium
- Verify that you are directed to composite checkout and that the plan product has been added to your cart correctly.
- Visit http://calypso.localhost:3000/checkout/YOURSITEHERE/domain-mapping:example.com?flags=composite-checkout-testing (except you can't use `example.com`; use a real domain but it doesn't have to be one that you intend to actually map).
- Verify that you are directed to composite checkout and that the domain mapping product has been added to your cart correctly for the domain you tried to map.
- Click to renew a purchased product (any kind) somewhere in calypso.
- Verify that the URL is of the form `/checkout/PRODUCT_SLUG/renew/PURCHASE_ID/YOURSITEHERE`.
- Verify that you are directed to _regular_ checkout and that the renewal product has been added to your cart correctly.

### CODE REVIEW

**Comment by sirbrillig** [comment]
I think I found a bug in this implementation; since `planSlug` is derived from `product` and state, `product` will be non-null and `planSlug` might be null when the callback is first called, causing the product lookup to fail and `canInitializeCart` to become true while we are still waiting for state to initialize.

**Comment by sirbrillig** [comment]
Ok, I believe I fixed it, although the client tests are failing with what appears to be a timeout. I suspect this is unrelated to this PR since our CI has been having problems today. I will continue to re-run the tests until I get a real error or they pass. In the mean time, this might as well be reviewed.

**Comment by sirbrillig** [comment]
I believe that should fix the tests.

**Comment by michaeldcain** [comment]
> and I had to use composite-testing-force instead of -testing to test the new Checkout.

After re-testing, this wasn't the case. Ignore me.

File: client/my-sites/checkout/checkout/composite-checkout.js [hunk 1]
```diff
@@ -768,34 +773,53 @@ function isNotCouponError( error ) {
...
-		cartMeta = planSlug.split( ':' )[ 1 ];
+	if ( productAlias.startsWith( 'domain-mapping:' ) && product_id ) {
+		debug( 'creating domain mapping product' );
+		cartMeta = productAlias.split( ':' )[ 1 ];
 		cartItem = domainMapping( { domain: cartMeta } );
+		cartItem.product_id = product_id;
 	}
 
-	if ( planSlug.startsWith( 'concierge-session' ) ) {
+	if ( productAlias.startsWith( 'concierge-session' ) ) {
```
[the remark this answers is not in the document]
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/composite-checkout.js]: I see your point; why not make this more clear? But I also know that there's likely to be several other issues with concierge-session products and I figure we might as well group those all together in their own PR. Those URLs won't make it here for now, so this logic (which was copied from regular checkout) is never used.

**Pull request merged**